### PR TITLE
[Impeller] GPU Tracer for GLES.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Trace GPU execution time on GLES using GL_EXT_disjoint_timer_query. This requires a per-app opt in from the Android Manifest with the key `"io.flutter.embedding.android.EnableOpenGLGPUTracing` set to true.

### CODE REVIEW

File: impeller/renderer/backend/gles/gpu_tracer_gles.cc [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+void GPUTracerGLES::ProcessQueries(const ProcTableGLES& gl) {
+  if (pending_traces_.empty()) {
+    return;
+  }
+
+  // For reasons unknown to me, querying the state of more than
+  // on query object per frame causes crashes on a Pixel 6 pro.
+  // It does not crash on an S10.
```
Comment (jason-simmons) [inline impeller/renderer/backend/gles/gpu_tracer_gles.cc:53]: Where exactly is the crash happening?

If the tracer can only consume one pending query per `MarkFrameStart` call, then that means the pending traces queue will grow unbounded each time a query result is unavailable at the next `MarkFrameStart`.

**Comment by jonahwilliams** [comment]
This actually works OK on the Pixel 7s we have in the device lab. On a Pixel 6 pro the crash is almost immediate and appears to be in the driver:

```
signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x0000000000000000
Cause: null pointer dereference
    x0  b40000797364acd0  x1  0000000000000000  x2  0000000000000000  x3  0000007999a60d98
    x4  0000000000000003  x5  0000000007ee634c  x6  71647164636d6471  x7  7f7f7f7f7f7f7f7f
    x8  0000000000000000  x9  0000000000000004  x10 0000000000000001  x11 0000000000000000
    x12 0000000000000008  x13 b400007b275cd128  x14 0000000000000007  x15 0000000000000028
    x16 0000000000000007  x17 0000000000000007  x18 000000798b636000  x19 b40000797364acd0
    x20 0000000000000000  x21 b4000079736b07b0  x22 b4000079735b4660  x23 b40000797360fef0
    x24 0000007999a65000  x25 000000799b889e6e  x26 0000007999a60d90  x27 0000000000008866
    x28 0000007999a60d98  x29 0000000000000001
    lr  00000079ece308a0  sp  0000007999a60a20  pc  00000079ece308ac  pst 0000000060001000
backtrace:
      #00 pc 000000000082d8ac  /vendor/lib64/egl/libGLES_mali.so (gles_fb_flush_compute_nx+92) (BuildId: 045075399765d011)
      #01 pc 00000000007e30ac  /vendor/lib64/egl/libGLES_mali.so (gles2_queryp_flush+108) (BuildId: 045075399765d011)
      #02 pc 00000000007e4050  /vendor/lib64/egl/libGLES_mali.so (gles_queryp_object_flush_for_result+64) (BuildId: 045075399765d011)
      #03 pc 00000000007e2760  /vendor/lib64/egl/libGLES_mali.so (gles2_queryp_get_query_object_getv+464) (BuildId: 045075399765d011)
      #04 pc 0000000002074c44  /data/app/~~KIK6odR8TVi-SXQaZ2msSQ==/com.gskinner.flutter.wonders-RurBsIIIDsouyhOiGFYG-Q==/base.apk!libflutter.so (offset 0xff23000) (auto impeller::GLProc<void (unsigned int, unsigned int, int*)>::operator()<unsigned int&, int, int*>(unsigned int&, int&&, int*&&) const+152) (BuildId: 46301208b8388191c9b7feb94f655a031d7f1eb7)
      #05 pc 000000000207803c  /data/app/~~KIK6odR8TVi-SXQaZ2msSQ==/com.gskinner.flutter.wonders-RurBsIIIDsouyhOiGFYG-Q==/ba
```
Based on some other reports, this seems like a fiddly API (https://bugs.chromium.org/p/chromium/issues/detail?id=453965) . I'm going to make usage opt in so we can use it on our benchmarks when necessary without worrying about broader device compatibility.

File: impeller/renderer/backend/gles/gpu_tracer_gles.cc [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  auto latest_query = pending_traces_.front();
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/renderer/backend/gles/gpu_tracer_gles.cc]: Done

File: impeller/renderer/backend/gles/gpu_tracer_gles.cc [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  raster_thread_ = std::this_thread::get_id();
+}
+
+void GPUTracerGLES::ProcessQueries(const ProcTableGLES& gl) {
+  if (pending_traces_.empty()) {
+    return;
+  }
+
+  // For reasons unknown to me, querying the state of more than
+  // on query object per frame causes crashes on a Pixel 6 pro.
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/renderer/backend/gles/gpu_tracer_gles.cc]: Done

File: impeller/renderer/backend/gles/gpu_tracer_gles.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  /// @brief Record the end of a frame workload.
+  void MarkFrameEnd(const ProcTableGLES& gl);
+
+ private:
+  void ProcessQueries(const ProcTableGLES& gl);
+
+  std::deque<uint32_t> pending_traces_;
+  std::optional<uint32_t> active_frame_ = std::nullopt;
+  std::thread::id raster_thread_;
+  bool has_started_frame_ = false;
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/renderer/backend/gles/gpu_tracer_gles.h]: Done

File: impeller/renderer/backend/gles/gpu_tracer_gles.cc [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  uint32_t query = 0;
+  gl.GenQueriesEXT(1, &query);
+  if (query == 0) {
+    return;
+  }
+
+  has_started_frame_ = true;
+
+  FML_DCHECK(!active_frame_.has_value());
+  FML_DCHECK(query != 0);
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/renderer/backend/gles/gpu_tracer_gles.cc]: Done

File: impeller/renderer/backend/gles/gpu_tracer_gles.cc [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    }
+    // Return the timer resolution in nanoseconds.
+    uint64_t duration = 0;
+    gl.GetQueryObjectui64vEXT(query, GL_QUERY_RESULT_EXT, &duration);
+    auto gpu_ms = duration / 1000000.0;
+
+    FML_TRACE_COUNTER("flutter", "GPUTracer",
+                      reinterpret_cast<int64_t>(this),  // Trace Counter ID
+                      "FrameTimeMS", gpu_ms);
+    FML_LOG(ERROR) << gpu_ms;
```
Comment (jonahwilliams) [inline impeller/renderer/backend/gles/gpu_tracer_gles.cc:71]: TODO: remove

File: impeller/renderer/backend/gles/gpu_tracer_gles.cc [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  // previous queries.
+  ProcessQueries(gl);
+
+  uint32_t query = 0;
+  gl.GenQueriesEXT(1, &query);
+  if (query == 0) {
+    return;
+  }
+
+  FML_DCHECK(!active_frame_.has_value());
```
Comment (jason-simmons) [inline impeller/renderer/backend/gles/gpu_tracer_gles.cc:35]: This duplicates the `active_frame_.has_value()` check above

File: impeller/renderer/backend/gles/gpu_tracer_gles.cc [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+                      reinterpret_cast<int64_t>(this),  // Trace Counter ID
+                      "FrameTimeMS", gpu_ms);
+    gl.DeleteQueriesEXT(1, &query);
+    pending_traces_.pop_front();
+  }
+}
+
+void GPUTracerGLES::MarkFrameEnd(const ProcTableGLES& gl) {
+  if (!enabled_ || std::this_thread::get_id() != raster_thread_ ||
+      !active_frame_.has_value() || !active_frame_.has_value()) {
```
Comment (jason-simmons) [inline impeller/renderer/backend/gles/gpu_tracer_gles.cc:78]: Remove the second copy of `!active_frame_.has_value()`

File: impeller/renderer/backend/gles/gpu_tracer_gles.cc (hunk 7 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/gles/gpu_tracer_gles.cc]: Done

File: impeller/renderer/backend/gles/gpu_tracer_gles.cc (hunk 8 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/gles/gpu_tracer_gles.cc]: Done

**Review by jason-simmons — APPROVED** [review]

**Pull request merged**